server/schedulers/adjacent_region.go

```diff
@@ -0,0 +1,292 @@
+		if len(args) == 2 {
+			leaderLimit, err := strconv.ParseUint(args[0], 10, 64)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			peerLimit, err := strconv.ParseUint(args[0], 10, 64)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			return newBalanceAdjacentRegionScheduler(opt, limiter, leaderLimit, peerLimit), nil
+		}
+		return newBalanceAdjacentRegionScheduler(opt, limiter), nil
+	})
+}
+
+type balanceAdjacentRegionScheduler struct {
+	*baseScheduler
+	opt          schedule.Options
+	selector     schedule.Selector
+	leaderLimit  uint64
+	peerLimit    uint64
+	lastKey      []byte
+	cacheRegions *adjacentState
+}
+
+type adjacentState struct {
+	assignedStoreIds []uint64
+	regions          []*core.RegionInfo
+	head             int
+}
+
+func (a *adjacentState) clear() {
+	a.assignedStoreIds = a.assignedStoreIds[:0]
+	a.regions = a.regions[:0]
+	a.head = 0
+}
+
+func (a *adjacentState) len() int {
+	return len(a.regions) - a.head
+}
+
+// newBalanceAdjacentRegionScheduler creates a scheduler that tends to disperse adjacent region
+// on each store.
+func newBalanceAdjacentRegionScheduler(opt schedule.Options, limiter *schedule.Limiter, args ...uint64) schedule.Scheduler {
+	filters := []schedule.Filter{
+		schedule.NewBlockFilter(),
+		schedule.NewStateFilter(opt),
+		schedule.NewHealthFilter(opt),
+		schedule.NewSnapshotCountFilter(opt),
+		schedule.NewStorageThresholdFilter(opt),
+	}
+	base := newBaseScheduler(limiter)
+	s := &balanceAdjacentRegionScheduler{
+		baseScheduler: base,
+		opt:           opt,
+		selector:      schedule.NewRandomSelector(filters),
+		leaderLimit:   defaultAdjacentLeaderLimit,
+		peerLimit:     defaultAdjacentPeerLimit,
+		lastKey:       []byte(""),
+	}
+	if len(args) == 2 {
+		s.leaderLimit = args[0]
+		s.peerLimit = args[1]
+	}
+	return s
+}
+
+func (l *balanceAdjacentRegionScheduler) GetName() string {
+	return "balance-adjacent-region-scheduler"
+}
+
+func (l *balanceAdjacentRegionScheduler) GetType() string {
+	return "adjacent-region"
+}
+
+func (l *balanceAdjacentRegionScheduler) GetMinInterval() time.Duration {
+	return minAdjacentSchedulerInterval
+}
+
+func (l *balanceAdjacentRegionScheduler) GetNextInterval(interval time.Duration) time.Duration {
+	return intervalGrow(interval, maxAdjacentSchedulerInterval, linearGrowth)
+}
+
+func (l *balanceAdjacentRegionScheduler) IsScheduleAllowed() bool {
+	return l.allowBalanceLeader() || l.allowBalanceLeader()
+}
+
+func (l *balanceAdjacentRegionScheduler) allowBalanceLeader() bool {
+	return l.limiter.OperatorCount(core.AdjacentLeaderKind) < l.leaderLimit
+}
+
+func (l *balanceAdjacentRegionScheduler) allowBalancePeer() bool {
+	return l.limiter.OperatorCount(core.AdjacentPeerKind) < l.peerLimit
+}
+
+func (l *balanceAdjacentRegionScheduler) Schedule(cluster schedule.Cluster) *schedule.Operator {
+	if l.cacheRegions == nil {
+		l.cacheRegions = &adjacentState{
+			assignedStoreIds: make([]uint64, 0, len(cluster.GetStores())),
```

Comment:
The stores should be updated everytime `Schedule` is called. If the namespace feature is enabled, the stores may change.

Reply:
the assigned store id is the exclude id.